components/sup/src/manager/mod.rs

```diff
@@ -778,37 +868,107 @@ impl Manager {
+        }
+
+        // Allow all existing futures to run to completion.
+        runtime
+            .shutdown_on_idle()
+            .wait()
+            .expect("Error waiting on Tokio runtime to shutdown");
+
+        release_process_lock(&self.fs_cfg);
+        self.butterfly.persist_data();
+
+        match shutdown_mode {
+            ShutdownMode::Normal | ShutdownMode::Restarting => Ok(()),
+            ShutdownMode::Departed => Err(sup_error!(Error::Departed)),
         }
     }
 
     fn check_for_updated_supervisor(&mut self) -> Option<PackageInstall> {
-        if let Some(ref mut updater) = self.self_updater {
-            return updater.updated();
+        if let Some(ref mut self_updater) = self.self_updater {
+            return self_updater.updated();
         }
         None
     }
 
-    /// Walk each service and check if it has an updated package installed via the Update Strategy.
-    /// This updates the Service to point to the new service struct, and then marks it for
-    /// restarting.
-    ///
-    /// The run loop's last updated census is a required parameter on this function to inform the
-    /// main loop that we, ourselves, updated the service counter when we updated ourselves.
-    fn check_for_updated_packages(&mut self) {
-        for service in self
+    /// Return the Services that currently have a newer package in
+    /// Builder. These are removed from the internal `services` vec
+    /// for further transformation into futures.
+    fn take_services_with_updates(&mut self) -> impl IntoIterator<Item = Service> {
+        let mut updater = self.updater.lock().expect("Updater lock poisoned");
+
+        let mut working_set = HashMap::new();
+        let mut state_services = self
             .state
             .services
             .write()
-            .expect("Services lock is poisoned!")
-            .values_mut()
-        {
-            if self
-                .updater
-                .check_for_updated_package(service, &self.census_ring, &self.launcher)
-            {
-                self.gossip_latest_service_rumor(&service);
-            }
-        }
+            .expect("Services lock is poisoned!");
+        mem::swap(state_services.deref_mut(), &mut working_set);
+
+        let (to_restart, mut no_action) =
+            working_set.drain().partition(|(current_ident, service)| {
+                match updater.check_for_updated_package(&service, &self.census_ring) {
+                    Some(new_package_ident) => {
+                        outputln!("Updating from {} to {}", current_ident, new_package_ident);
+                        true
+                    }
+                    None => {
+                        trace!("No update found for {}", current_ident);
+                        false
+                    }
+                }
+            });
+        mem::swap(&mut no_action, state_services.deref_mut());
+
+        to_restart.into_iter().map(|(_ident, service)| service)
+    }
```

Comment:
Wow, thanks for the very thorough analysis! I agree that your alternative is better from an allocation standpoint. I'm gonna use that one instead for that reason, and because it's probably a bit easier to understand than what's there right now.

👍 👍 👍